kimchi/src/circuits/gate.rs

```diff
@@ -114,6 +116,279 @@ pub enum GateType {
+            Gate::RangeCheck0(_) => GateType::RangeCheck0,
+            Gate::RangeCheck1(_) => GateType::RangeCheck1,
+            Gate::ForeignFieldAdd(_) => GateType::ForeignFieldAdd,
+            Gate::ForeignFieldMul(_) => GateType::ForeignFieldMul,
+            Gate::Xor16(_) => GateType::Xor16,
+            Gate::Rot64(_) => GateType::Rot64,
+        }
+    }
+
+    pub fn domain(&self) -> Domain {
+        // TODO: In fucture this could be computed from degree of gate constraint
+        // TODO: Check these over to see if any can be reduced
+        match *self {
+            Gate::Generic(_) => Domain::D4,
+            Gate::Poseidon(_) => Domain::D8,
+            Gate::CompleteAdd(_) => Domain::D4,
+            Gate::VarBaseMul(_) => Domain::D8,
+            Gate::EndoMul(_) => Domain::D8,
+            Gate::EndoMulScalar(_) => Domain::D8,
+            Gate::CairoClaim(_) => Domain::D8, // TODO: set Cairo gate domains correctly
+            Gate::CairoInstruction(_) => Domain::D8,
+            Gate::CairoFlags(_) => Domain::D8,
+            Gate::CairoTransition(_) => Domain::D8,
+            Gate::RangeCheck0(_) => Domain::D8,
+            Gate::RangeCheck1(_) => Domain::D8,
+            Gate::ForeignFieldAdd(_) => Domain::D8,
+            Gate::ForeignFieldMul(_) => Domain::D8,
+            Gate::Xor16(_) => Domain::D8,
+            Gate::Rot64(_) => Domain::D8,
+        }
+    }
+
+    pub fn argument_type(&self) -> ArgumentType {
+        match *self {
+            Gate::Generic(_) => generic::Generic::<F>::ARGUMENT_TYPE,
+            Gate::Poseidon(_) => poseidon::Poseidon::<F>::ARGUMENT_TYPE,
+            Gate::CompleteAdd(_) => complete_add::CompleteAdd::<F>::ARGUMENT_TYPE,
+            Gate::VarBaseMul(_) => varbasemul::VarbaseMul::<F>::ARGUMENT_TYPE,
+            Gate::EndoMul(_) => endosclmul::EndosclMul::<F>::ARGUMENT_TYPE,
+            Gate::EndoMulScalar(_) => endomul_scalar::EndomulScalar::<F>::ARGUMENT_TYPE,
+            Gate::CairoClaim(_) => turshi::Claim::<F>::ARGUMENT_TYPE,
+            Gate::CairoFlags(_) => turshi::Flags::<F>::ARGUMENT_TYPE,
+            Gate::CairoTransition(_) => turshi::Transition::<F>::ARGUMENT_TYPE,
+            Gate::CairoInstruction(_) => turshi::Instruction::<F>::ARGUMENT_TYPE,
+            Gate::RangeCheck0(_) => range_check::circuitgates::RangeCheck0::<F>::ARGUMENT_TYPE,
+            Gate::RangeCheck1(_) => range_check::circuitgates::RangeCheck1::<F>::ARGUMENT_TYPE,
+            Gate::ForeignFieldAdd(_) => {
+                foreign_field_add::circuitgates::ForeignFieldAdd::<F>::ARGUMENT_TYPE
+            }
+            Gate::ForeignFieldMul(_) => {
+                foreign_field_mul::circuitgates::ForeignFieldMul::<F>::ARGUMENT_TYPE
+            }
+            Gate::Xor16(_) => xor::Xor16::<F>::ARGUMENT_TYPE,
+            Gate::Rot64(_) => rot::Rot64::<F>::ARGUMENT_TYPE,
+        }
+    }
+
+    pub fn constraint_count(&self) -> u32 {
```

Comment:
Maybe create a function that returns what's the maximum constraint count. Useful for exponents inside linearization.

Reply:
That's in iteration2 PR